### PR TITLE
Fix builtin linkage

### PR DESCRIPTION
**Pull request by qjerome**
- fixing memset return value for proper BTF validation
- fixing memset + memcpy linkage type when loading BPF program

### CODE REVIEW

File: bpf/aya-bpf/src/lib.rs [hunk 1]
```diff
@@ -57,11 +57,12 @@ pub trait BpfContext {
 }
 
 #[no_mangle]
-pub unsafe extern "C" fn memset(s: *mut u8, c: c_int, n: usize) {
+pub unsafe extern "C" fn memset(s: *mut u8, c: c_int, n: usize) -> u8 {
```
Comment (alessandrod) [inline bpf/aya-bpf/src/lib.rs:60]: This is not needed now that we're making these functions static I think?

We should still fix our memory builtins to 1) be faster 2) work as globals, but we should do that in a separate PR and for all of them

File: aya-obj/src/btf/btf.rs [hunk 2]
```diff
@@ -536,22 +536,30 @@ impl Btf {
...
-                        fixed_ty.set_linkage(FuncLinkage::Static);
+                BtfType::Func(ty) => {
+                    let name = self.string_at(ty.name_offset)?;
+                    // Sanitize FUNC
+                    if !features.btf_func {
+                        debug!("{}: not supported. replacing with TYPEDEF", kind);
+                        let typedef_type =
+                            BtfType::Typedef(Typedef::new(ty.name_offset, ty.btf_type));
+                        types.types[i] = typedef_type;
+                    } else if !features.btf_func_global || name == "memset" || name == "memcpy" {
```
Comment (alessandrod) [inline aya-obj/src/btf/btf.rs:547]: this needs to be done for memmove and memcmp too

**Review by alessandrod — CHANGES_REQUESTED** [review]
Thanks for working on this! See comments

File: bpf/aya-bpf/src/lib.rs (hunk 1 above)
Comment (qjerome) [reply bpf/aya-bpf/src/lib.rs:60]: You are right ! It works without. I will change that

File: aya-obj/src/btf/btf.rs (hunk 2 above)
Comment (qjerome) [reply aya-obj/src/btf/btf.rs]: Done

File: bpf/aya-bpf/src/lib.rs (hunk 1 above)
Comment (qjerome) [reply bpf/aya-bpf/src/lib.rs]: Done

File: aya-obj/src/btf/btf.rs [hunk 3]
```diff
@@ -1350,6 +1363,61 @@ mod tests {
...
+            },
+            BtfParam {
+                name_offset: btf.add_string("b".to_string()),
+                btf_type: int_type_id,
+            },
+        ];
+        let func_proto_type_id =
+            btf.add_type(BtfType::FuncProto(FuncProto::new(params, int_type_id)));
+
+        ["memset", "memcpy", "memcmp", "memmove"]
```
Comment (alessandrod) [inline aya-obj/src/btf/btf.rs:1390]: Please make this a for loop

File: aya-obj/src/btf/btf.rs [hunk 4]
```diff
@@ -536,22 +536,35 @@ impl Btf {
...
+                        let typedef_type =
+                            BtfType::Typedef(Typedef::new(ty.name_offset, ty.btf_type));
+                        types.types[i] = typedef_type;
+                    } else if !features.btf_func_global
+                        || name == "memset"
+                        || name == "memcpy"
+                        || name == "memmove"
+                        || name == "memcmp"
+                    {
+                        // Sanitize BTF_FUNC_GLOBAL and memset, memcpy
```
Comment (alessandrod) [inline aya-obj/src/btf/btf.rs:553]: The comment here should say:

```
Sanitize BTF_FUNC_GLOBAL when not supported and ensure that memory builtins are marked as static. Globals are type checked and verified separately from their callers, while instead we want tracking info (eg bound checks) to be propagated to the memory builtins.  
```

**Review by alessandrod — COMMENTED** [review]
Thanks, almost there!

File: aya-obj/src/btf/btf.rs (hunk 4 above)
Comment (qjerome) [reply aya-obj/src/btf/btf.rs]: Done

File: aya-obj/src/btf/btf.rs (hunk 3 above)
Comment (qjerome) [reply aya-obj/src/btf/btf.rs]: Done
